bindings/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,343 @@
+)
+
+var _ bindings.OutputBinding = &Clickhouse{}
+
+type Clickhouse struct {
+	db     *sql.DB
+	logger logger.Logger
+}
+
+func NewClickhouse(logger logger.Logger) *Clickhouse {
+	return &Clickhouse{logger: logger}
+}
+
+func (c *Clickhouse) Init(metadata bindings.Metadata) error {
+	p := metadata.Properties
+	url, ok := p[connectionURLKey]
+	if !ok || url == "" {
+		return ErrMetadataURLKeyMissing
+	}
+
+	db, err := initDB(url, metadata.Properties[pemPathKey])
+	if err != nil {
+		return err
+	}
+
+	if err = propertyToInt(p, maxIdleConnsKey, db.SetMaxIdleConns); err != nil {
+		return err
+	}
+
+	if err = propertyToInt(p, maxOpenConnsKey, db.SetMaxOpenConns); err != nil {
+		return err
+	}
+
+	if err = propertyToDuration(p, connMaxIdleTimeKey, db.SetConnMaxIdleTime); err != nil {
+		return err
+	}
+
+	if err = propertyToDuration(p, connMaxLifetimeKey, db.SetConnMaxLifetime); err != nil {
+		return err
+	}
+
+	c.db = db
+	return nil
+}
+
+func (c *Clickhouse) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	if req == nil {
+		return nil, ErrRequestRequired
+	}
+
+	if req.Operation == closeOperation {
+		return nil, c.db.Close()
+	}
+
+	if req.Metadata == nil {
+		return nil, ErrMetadataRequired
+	}
+	c.logger.Debugf("operation: %v", req.Operation)
+
+	s, ok := req.Metadata[commandSQLKey]
+	if !ok || s == "" {
+		return nil, ErrCommandSQLKeyRequired
+	}
+
+	startTime := time.Now().UTC()
+
+	resp := &bindings.InvokeResponse{
+		Metadata: map[string]string{
+			respOpKey:        string(req.Operation),
+			respSQLKey:       s,
+			respStartTimeKey: startTime.Format(time.RFC3339Nano),
+		},
+	}
+
+	switch req.Operation {
+	case execOperation:
+		if err := c.exec(s); err != nil {
+			return nil, err
+		}
+
+	case insertOperation, updateOperation, deleteOperation:
+		if req.Data == nil {
+			return nil, ErrDataRequired
+		}
+		var vals []interface{}
+		if err := msgpack.Unmarshal(req.Data, &vals); err != nil {
+			return nil, err
+		}
+		if err := c.execByTransaction(s, vals...); err != nil {
+			return nil, err
+		}
+	case queryOperation:
+		d, err := c.query(s)
+		if err != nil {
+			return nil, err
+		}
+		resp.Data = d
+
+	default:
+		return nil, ErrInvalidOperation
+	}
+
+	endTime := time.Now().UTC()
+	resp.Metadata[respEndTimeKey] = endTime.Format(time.RFC3339Nano)
+	resp.Metadata[respDurationKey] = endTime.Sub(startTime).String()
+
+	return resp, nil
+}
+
+func (c Clickhouse) Operations() []bindings.OperationKind {
+	return []bindings.OperationKind{execOperation, queryOperation, closeOperation, insertOperation, updateOperation, deleteOperation}
+}
+
+func (c *Clickhouse) Close() error {
+	if c.db != nil {
+		return c.db.Close()
+	}
+	return nil
+}
+
+func (c *Clickhouse) query(sql string) ([]byte, error) {
+	rows, err := c.db.Query(sql)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error executing %s", sql)
+	}
+
+	defer func() {
+		_ = rows.Close()
+		_ = rows.Err()
```

Comment:
What it is for?

Reply:
@uptutu why do you even have these assignments? And Why do you need to call the error function after closing the Query results?